Preload sticker manifests

sticker201 opened the pull request:
This addresses #47 by implementing a webpack plugin that fetches all known pack manifests at build time. This PR contains the changes in #32.

You can take a look at https://sticker201.github.io/signalstickers/

The output of the build phase is [stickerData.json](https://sticker201.github.io/signalstickers/stickerData.json) which contains the info from `stickers.json` as well as the manifest record for each pack. This is now loaded instead of `stickers.json` by the client.

sticker201 commented on src/plugins/FetchStickerDataPlugin.ts:
```diff
@@ -0,0 +1,130 @@
+ * Decrypts a manifest returned from the Signal API using a sticker pack's
+ * pack key, provided from stickers.json.
+ */
+async function decryptManifest(encodedKey: string, rawManifest: any) {
```
Unfortunately, we can't use the logic in `crypto.ts` to do this work as it's based on `window.crypto` which isn't available server side during the build process.

sticker201 commented on src/lib/stickers.ts:
```diff
@@ -213,7 +213,7 @@ export async function getEmojiForSticker(id: string, key: string, stickerId: num
  */
 export function fuzzySearchStickerPacks(needle: string, haystack: Array<StickerPack>): Array<StickerPack> {
-  const searchKeys = ['manifest.title', 'manifest.author', 'meta.tags', 'meta.source'];
+  const searchKeys = ['manifest.title', 'manifest.author', 'meta.tags'];
```
Fuzzy search started to break down with a lot of packs. In general, the main issue was any long string tended to have a lot of false positives in the fuzzy filter. I've mitigated this for now by:
* no longer filtering on source (it tended to only produce unexpected results)
* breaking the tags key up into an array, so each tag is searched individually rather than the full list of tags as one string

romainricard replied:
This is a great! :+1:

romainricard commented:
It looks like it works great! 

One point tho: for each pack, you add its whole manifest in [stickerData.json](https://sticker201.github.io/signalstickers/stickerData.json), including the emoji mapping. Even if it decreases code complexity, it multiplies the weight of the generated file by 3, by including a majority of mappings the user won't need (1,6M with mappings vs. 552K without).
How would you feel about not adding the emoji mapping in `stickerData.json`, and (re)loading it in `/pack/` for each item viewed?

sticker201 commented:
> One point tho: for each pack, you add its whole manifest in [stickerData.json](https://sticker201.github.io/signalstickers/stickerData.json), including the emoji mapping. Even if it decreases code complexity, it multiplies the weight of the generated file by 3, by including a majority of mappings the user won't need (1,6M with mappings vs. 552K without).
> How would you feel about not adding the emoji mapping in `stickerData.json`, and (re)loading it in `/pack/` for each item viewed?

@romainricard I've gone ahead and done as you requested. `stickerData.json` no longer contains the `stickers` key of the manifest. When you tap a pack we'll fetch the manifest.

I also made a couple other small fixes:
* Removed packs that were duplicates (different ID, but had been ported from telegram more than once)
* Added retry / delay logic in the build phase around the manifest. Sometimes the Signal service seems to get mad when we fetch a bunch of manifests at one and start timing out. This ensures the build always succeeds.

sticker201 commented:
I also addressed #51 and converted `stickers.json` to yml as well as made the appropriate changes on the contribution page.

sticker201 commented:
I think this should be good to go unless you have any other feedback @romainricard !

sticker201 commented:
> Working on a way to get builds/checks to run on forked repositories. Let's hold off on merging this until we can get that solved.

Let me know if there's anything else I can do to help push this forward. FWIW, with github actions the tests run on the fork. It's a step of disconnect to check things, but you can see results for the last commit in this branch for example here:
https://github.com/sticker201/signalstickers/commit/c07be0132ad74a08f3df42b57d1744a485217346/checks?check_suite_id=395846657

darkobits commented:
> Let me know if there's anything else I can do to help push this forward. FWIW, with github actions the tests run on the fork. It's a step of disconnect to check things, but you can see results for the last commit in this branch for example here:

Correct, the check(s) run in the context of the forked repo, but don't have access to any custom secrets and its `GITHUB_TOKEN` (which we don't use) is ready-only. On top of this, the _status_ of the check is not getting reported back to this PR, as you can see from the `Expected — Waiting for status to be reported` descriptor.

sticker201 commented:
> On top of this, the _status_ of the check is not getting reported back to this PR, as you can see from the `Expected — Waiting for status to be reported` descriptor.

Right, my point was just that I'm not sure this needs to block this PR since the tests and build have run fine and upon merge to master the build and publish will happen as expected. I agree it should be fixed so that we can see the test status passed in PRs, but that seems like an independent issue.

darkobits commented:
I'm getting some errors trying to start and/or build this branch locally, which is why I mentioned it (Node `12.13.1`, NPM `6.13.6`).

I just created a new branch, `develop`, that is currently up-to-date with `master`. Perhaps we can merge this into `develop`, get the checks passing there, then merge to `master`. If that sounds good to you, you can change the base branch to `develop`, which should enable auto-merging.

---

In the meantime, we can further discuss what to do about potentially migrating to Travis or another CI solution with @romainricard.

sticker201 commented:
Okay, I did add some new dependencies so you'll have to `npm install`, but after that just running `npm start` works for me locally. If you share the errors I can try and help debug.

romainricard commented:
I've no problem migrating to Travis, as it is free for open-source project! 

On a side note, I thought about https://github.com/romainricard/signalstickers/pull/32 recently. If we bring the Telegram stickers with this PR, it could be a good thing to allow to filter Telegram/_Signal exclusive_ stickers. I would be more in favor of having only _exclusive_ stickers by default, so the user won't feel overwhelmed by 1000+ stickers, do you have an opinion @sticker201 @darkobits ?

darkobits commented:
> I've no problem migrating to Travis, as it is free for open-source project!

Awesome! Happy to provide any assistance I can in getting the integration set up. Again, I would really like to have passing checks in place on all PRs as soon as possible to avoid inadvertently merging changes that would break the site.

> On a side note, I thought about #32 recently. If we bring the Telegram stickers with this PR, it could be a good thing to allow to filter Telegram/Signal exclusive stickers. I would be more in favor of having only exclusive stickers by default, so the user won't feel overwhelmed by 1000+ stickers, do you have an opinion

I've been doing some work with this branch, and have definitely noticed that the addition of something like 2,000-ish sticker packs has created some issues that will need to be addressed in the very near future.

The biggest issue I'm seeing so far has to do with quality and potential abuse. I'm seeing a lot of sticker packs with names that start with non-alphabetic characters like `@`, which do nothing to add semantic value to the pack's name, but do ensure that the pack is moved to the top of the list in any case where alpha-sorting is being used. I see this as a potential form of abuse that should be mitigated. More generally, low-quality sticker packs add noise to search results that make it more difficult for the user to find what they want.

I think a reasonable short-term option would be that we require all PRs for sticker pack submissions include a link to the sticker pack preview page on signalstickers.com so that the reviewer can easily review the pack's content and request that the PR author re-create the sticker pack if there are any issues, such as poor naming standards, copyright violations, duplicate pack submissions, and so forth. GitHub's [issue templates](https://help.github.com/en/github/building-a-strong-community/about-issue-and-pull-request-templates#issue-templates) would be useful here to make this requirement clear to PR authors.

---

With respect to the `exclusive` attribute, this would need to be another field we add to the YAML spec and the Contribute page and will rely entirely on the honesty of the user submitting the sticker pack. If we decide to promote/demote search results based on this flag, that means it becomes another potential vector for abuse. I see the massive Telegram import as a one-time thing, and in the future we won't really have a good way to know or find out if any given submission is a Signal exclusive or not.

One alternative we might consider is adding a `Telegram` tag to all the packs in the Telegram PR, which would allow users to search for "telegram" and get a mostly accurate result set that way.

darkobits commented:
Going to go ahead and merge this to `develop`, may open a PR against that branch if needed.

sticker201 commented:
> One alternative we might consider is adding a `Telegram` tag to all the packs in the Telegram PR, which would allow users to search for "telegram" and get a mostly accurate result set that way.

This is already done.

> The biggest issue I'm seeing so far has to do with quality and potential abuse. I'm seeing a lot of sticker packs with names that start with non-alphabetic characters like `@`, which do nothing to add semantic value to the pack's name, but do ensure that the pack is moved to the top of the list in any case where alpha-sorting is being used. I see this as a potential form of abuse that should be mitigated. More generally, low-quality sticker packs add noise to search results that make it more difficult for the user to find what they want.

I think probably the answer here is to stop alpha sorting. The ideal sorting IMO would be by popularity. We would have to add some tracking to count clicks on the "Add to signal" button per pack which probably means a database of some sort, though we might be able to get away with something simpler (though we'd want some protection to prevent people gaming the system by tapping the link a bunch).

> I see the massive Telegram import as a one-time thing, and in the future we won't really have a good way to know or find out if any given submission is a Signal exclusive or not.

I agree with this sentiment. I'd also point out that a large majority of the packs that are already on the site are ported from Telegram. There are only a dozen or so original packs AFAIK.

> I think a reasonable short-term option would be that we require all PRs for sticker pack submissions include a link to the sticker pack preview page on signalstickers.com so that the reviewer can easily review the pack's content and request that the PR author re-create the sticker pack if there are any issues, such as poor naming standards, copyright violations, duplicate pack submissions, and so forth. GitHub's [issue templates](https://help.github.com/en/github/building-a-strong-community/about-issue-and-pull-request-templates#issue-templates) would be useful here to make this requirement clear to PR authors.

Alternatively / in addition to this we could introduce an optional `alias` field to the yml that lets us override the name on a pack-by-pack basis. So we wouldn't have to have all names in the `stickers.yml`, but if there's a problematically named pack (like `(my) Best of Messenger`) we could assign an alias like `Best of Messenger` that overrides the `title` in the manifest.

sticker201 commented:
One other thing we could do is curate "featured" packs and add a `featured` key to the `stickers.yml` where we hand select the packs we think are high quality and that we want to show on initial load. I'd strongly prefer we do something based on actual popularity though.
